### ISSUE TITLE
Better customization info and command behavior

### ISSUE THREAD

**Issue by dmolina** [weirdNox/org-noter#7]
I suggest you to be more explicit in the README about customization, in particular how the position of division can be selected. Actually, I am interested in dividing the screen vertically (sometimes I work in the screen of a small laptop) instead of horizontally, and I am not sure how it can be done (reading the code, it seems that it is customized, but I am not sure how to do it).

**Comment by weirdNox** [comment]
Ok, I'll improve the [customization section](https://github.com/weirdNox/org-noter#customization-), then. I'll make a short summary here, and then, if you understand, I'll adapt it to the README. 

There are two kinds of customizations:

1. The ones you make globally, affecting every session
2. The ones specific to a document itself, which override the global setting

The global settings are changed with either the [customization interface from emacs](https://www.gnu.org/software/emacs/manual/html_node/emacs/Easy-Customization.html#Easy-Customization) or directly in your init file. To find which settings are available, you may use the customization interface or you may just read the source code.

The best way to set document-specific settings is by using the utility commands I talk about in the README. In order to use the utility commands, you need an open a session, and after that you just run them to change the settings for that session only or permanently. You could also change the properties of the headline directly, but you'd have to know which are available and what is the settings format.

I think one of the problems you are encountering is that when you run `org-noter-set-notes-window-location` command, you don't see any effect; the only thing it does is to change the setting itself. For it to have any effect, you'd need to close the current window, and even then, it would not work for what you want because of a [TODO I have](https://github.com/weirdNox/org-noter/blob/5fa69bac578493041f516aa781a6819c1c218130/org-noter.el#L400) :P 

The vertical setting, *currently*, only works when starting a new session with that setting. I'll try to fix it now, and I'll tell you when it's done for you to test.

So, in summary, a workaround for what you want is to run `org-noter-set-notes-window-location` with a prefix argument (eg. `C-u`) so it is persistent, and then *restart* the session. 

Hope this helps you understand! :)

**Comment by weirdNox** [comment]
Oh, another side note: if you just want to change the notes window location temporarily, you can do whatever you want with it. If you delete the one on the session frame and then open it on another frame, for example, it will be the one that is used. 

So, if you don't want to set it persistently on the document, you could just change it manually and open the window below using Emacs window commands.

**Comment by weirdNox** [comment]
I just submitted the fix I was talking about in 9fcc27650d001638d9636a0667c8b6bc4be2ab0f, so `org-noter-set-notes-window-location` has immediate effects, and reworded the README in 5dc976999640a30d023afb5de44000e2105ccc05. It will take some time until it is on MELPA, but once it is, give it a try! :)

**Comment by stardiviner** [comment]
@weirdNox Can `org-noter` smartly on decide split horizontal or veritical based on Emacs frame and window width?
I have some elisp code about get those info:
```org
** get display monitor attributes

Additionally, if you are using the development version (to be released as 24.4):

Multi-monitor support has been added.

New functions ~display-monitor-attributes-list~ and ~frame-monitor-attributes~ can
be used to obtain information about each physical monitor on multi-monitor
setups.

#+begin_src emacs-lisp :results verbatim
(display-monitor-attributes-list)
#+end_src

#+RESULTS:
: (((name . "HDMI1") (geometry 0 0 1920 1080) (workarea 0 30 1920 1050) (mm-size 600 340) (frames #<frame Emacs λ Hacking 0x13dada0>) (source . "Gdk")))

#+begin_src emacs-lisp :results verbatim
(x-display-monitor-attributes-list)
#+end_src

#+RESULTS:
: (((name . "HDMI1") (geometry 0 0 1920 1080) (workarea 0 30 1920 1050) (mm-size 600 340) (frames #<frame Emacs λ Hacking 0x13dada0>) (source . "Gdk")))

#+begin_src emacs-lisp :results verbatim
(frame-monitor-attributes)
#+end_src

#+RESULTS:
: ((name . HDMI1) (geometry 0 0 1920 1080) (workarea 0 30 1920 1050) (mm-size 600 340) (frames #<frame Emacs λ Hacking 0x13dada0>) (source . Gdk))

** get screen size: width & height
:PROPERTIES:
:URL:      https://stackoverflow.com/questions/16481984/get-width-of-current-monitor-in-emacs-lisp
:END:

*** use function ~display-monitor-attributes-list~

#+begin_src emacs-lisp :results output
(print (format "pixels: %s" (cdddr (car (cdr (car (display-monitor-attributes-list)))))))
(print (format "length: %s" (/ (car (cdddr (car (cdr (car (display-monitor-attributes-list)))))) 10)))
#+end_src

#+RESULTS:
: 
: "pixels: (1366 768)"
: 
: "length: 136"

*** use function ~frame-monitor-attributes~

#+begin_src emacs-lisp :results verbatim
(cdddr (car (cdr (frame-monitor-attributes))))
#+end_src

#+RESULTS:
: (1920 1080)

*** display-pixel functions -- ~display-pixel-[width|height]~

#+begin_src emacs-lisp
(prin1
 (format "width x height: %s x %s"
         (display-pixel-width)
         (display-pixel-height)))
#+end_src

#+RESULTS:
: width x height: 1920 x 1080

*** get frame width & height -- ~frame-pixel-[width|height]~

#+begin_src emacs-lisp
(prin1 (format "frame width x height: %s %s" (frame-pixel-width) (frame-pixel-height)))
#+end_src

#+RESULTS:
: frame width x height: 1920 1023

```

**Comment by stardiviner** [comment]
The decide strategy can be like:
- If Emacs frame (or The Display Monitor size) is much larger than Emacs window, then has space to split window horizontal. If it is a small screen, split window vertical is better.
Why I want it be automatic, and smartly?
Because I often connect my small screen (12') laptop with Big monitor (27'). With this automatic will be great.

**Comment by weirdNox** [comment]
I like the idea of smartly choosing the split direction. However could you explain better the suggested strategy? What do you mean mean with Emacs window?

**Comment by stardiviner** [comment]
That's great you like this idea.
About the strategy:
Let me declare some concepts:
- Display monitor -> Emacs frame -> Emacs window (A window in Emacs not the Emacs application window, sorry for my unclear point)
The core of my strategy is "comparing size of current window with others".
(I think that it will be great if this deciding strategy can be a option to function like `(defcustom `org-noter-split-strategy 'org-noter-split-window-smartly)`, of course name is you decide, I just give an example. So user can define their own stategy. Just like `display-buffer-alist` mechanism.)
Because I often switch between different Display monitors, they have different screen size.
So when I open `org-noter`, it compare current window width (code is in upper post) with the current frame size.
- If it is larger than "half of frame width" (usually user will split window in frame horizontally), then split org-noter document horizontally.
  - if smaller than "half of frame width", then split vertically.
  - if it is too small under a minium size (width and height), then just popup a new frame with current org-mode buffer and the org-noter document side by side. Put a small window beside is meanless for viewing.

What do you think?

**Comment by weirdNox** [comment]
So, if I'm understanding correctly, the test you describe should be done when creating a session, testing the windows you already have created, which would then set the correct strategy for the rest of the session, right? Using you strategy, what happens when the user only has one window occupying the full frame?

> (I think that it will be great if this deciding strategy can be a option to function like (defcustom org-noter-split-strategy 'org-noter-split-window-smartly), of course name is you decide, I just give an example. So user can define their own stategy. Just like display-buffer-alist mechanism.)

Yes, I was thinking about that!

I'm sorry if I'm not understanding correctly what you mean! It's just that I like having no nuances when adding new features! Bear with me :P

**Comment by weirdNox** [comment]
Just thought about something: if you are able to write the logic you want in elisp, then you could use an advice to show what you mean.

```elisp
(advice-add
   'org-noter--create-session :around
   (lambda (orig-function &rest args)
     (let ((org-noter-notes-window-location 'other-frame))
       (apply orig-function args))))
```

For example, this would change the default option to `other-frame` unconditionally, but you could add some logic in there so it would do what you want!

**Comment by stardiviner** [comment]
Oh, I totally misunderstand current `org-noter` strategy. Sigh. It currently open a session in a new frame by default. I was thought it will just split PDF document at side (horizontal or vertically) based on `org-noter-notes-window-location` (without new frame). After checked out source code, it will make a new frame no matter what `org-noter-notes-window-location` is. Sorry my fault. I upgrade `org-noter` but have not used yet recently. My strategy idea is based on current frame.
BTW, Can I have a session just in current frame? I always set shortcuts for programs like Emacs, Firefox, IM etc under Linux KDE. But a new frame switching is not fast as I want.

**Comment by weirdNox** [comment]
I just added the possibility of having a session on the selected frame, if you still want it :)

And ok, I'll leave it like this for now. However, if you want to try having smart splitting once again and if you know how to write some elisp, you don't even need to wait for my decision. Using the advice I described above, I think you have everything you need. You may submit it if you think other people may like it!

**Comment by stardiviner** [comment]
@weirdNox Ok, I will try to write a basic function, then inject it with advice. I will post it here.

**Comment by stardiviner** [comment]
Here is my strategy code prototype:
```elisp
(defun display-pixel-offset-calculate ()
  "Calculate the offset pixels between Emacs window, frame, and displayer."
  (let (pixel-offset)
    (winner-remember)
    (set-frame-parameter nil 'fullscreen 'maximized)
    (delete-other-windows) (split-window-horizontally)
    (setq pixel-offset (- (display-pixel-width) (* (window-width (frame-selected-window) t) 2)))
    (winner-undo) (winner-undo) ; restore windows configurations.
    (toggle-frame-maximized)            ; back to original size.
    pixel-offset))

(defun split-window-strategy ()
  (let* ((offset (display-pixel-offset-calculate))
         (window-width (+ (window-width (frame-selected-window) t) offset))) ; 660 + (offset: 30~48 on my laptop)
    (if (> window-width (/ (display-pixel-width) 2)) ; 683
        (split-window-horizontally)
      (if (< window-width (/ (frame-pixel-width) 2)) ; 681
          (split-window-vertically)
        ))))
```

**Comment by weirdNox** [comment]
Hello @stardiviner, is the advice you made working OK? As this is a very specific thing, if you are happy with the solution you arrived at, maybe I'll just close this issue :)

**Comment by stardiviner** [comment]
@weirdNox yeah, I have not dig into this solution for a long time. almost forgotten. I will update any news on this issue if have. Can be closed. Thanks for your attention.

**Comment by weirdNox** [comment]
Ok, thanks :)
